Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -435,4 +454,191 @@ theorem weierstrassMod_zero_left [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
+    have hp1 : p.1 < n := (Finset.antidiagonal.fst_le hp).lt_of_ne <| by
+      contrapose! hnmem
+      rwa [Finset.mem_singleton, Finset.antidiagonal_congr hp (by simp)]
+    rw [coeff_of_lt_order p.1 (by
+      rwa [← ENat.lt_lift_iff (h := order_finite_iff_ne_zero.2 hg),
+        ENat.lift_eq_toNat_of_lt_top]), zero_mul]), Finset.sum_singleton,
+    coeff_map, coeff_map, coeff_zero_eq_constantCoeff, mul_comm, Eq.comm] at H2
+  rw [isUnit_iff_constantCoeff, ← isUnit_map_iff (IsLocalRing.residue A)]
+  exact isUnit_of_mul_eq_one _ _ H2
+
+theorem IsWeierstrassDivision.isWeierstrassFactorization
+    {g q : A⟦X⟧} {r : A[X]} (hg : g.map (IsLocalRing.residue A) ≠ 0)
+    (H : (X ^ (g.map (IsLocalRing.residue A)).order.toNat).IsWeierstrassDivision g q r) :
+    g.IsWeierstrassFactorization
+      (Polynomial.X ^ (g.map (IsLocalRing.residue A)).order.toNat - r)
+      ↑(H.isUnit_of_map_ne_zero hg).unit⁻¹ := by
+  have H1 : r.degree < (g.map (IsLocalRing.residue A)).order.toNat := H.1
+  set n := (g.map (IsLocalRing.residue A)).order.toNat
+  set f := Polynomial.X ^ n - r
+  replace H1 : r.degree < (Polynomial.X (R := A) ^ n).degree := by rwa [Polynomial.degree_X_pow]
+  have hfdeg : f.natDegree = n := by
+    suffices f.degree = n by rw [Polynomial.natDegree, this]; rfl
+    rw [Polynomial.degree_sub_eq_left_of_degree_lt H1, Polynomial.degree_X_pow]
+  refine ⟨⟨⟨fun {i} hi ↦ ?_⟩, .sub_of_left (Polynomial.monic_X_pow _) H1⟩, Units.isUnit _, ?_⟩
+  · rw [hfdeg] at hi
+    simp_rw [f, Polynomial.coeff_sub, Polynomial.coeff_X_pow, if_neg hi.ne, zero_sub, neg_mem_iff]
+    have := H.coeff_f_sub_r_mem hi
+    rwa [map_sub, coeff_X_pow, if_neg hi.ne, zero_sub, neg_mem_iff, Polynomial.coeff_coe] at this
+  · have := congr($(H.2) * ↑(H.isUnit_of_map_ne_zero hg).unit⁻¹)
+    rw [add_mul, mul_assoc, IsUnit.mul_val_inv, mul_one, ← sub_eq_iff_eq_add] at this
+    simp_rw [← this, f, Polynomial.coe_sub, Polynomial.coe_pow, Polynomial.coe_X, sub_mul]
+
+theorem IsWeierstrassFactorization.isWeierstrassDivision
+    {g : A⟦X⟧} {f : A[X]} {h : A⟦X⟧} (H : g.IsWeierstrassFactorization f h) :
+    (X ^ (g.map (IsLocalRing.residue A)).order.toNat).IsWeierstrassDivision g ↑H.isUnit.unit⁻¹
+      (Polynomial.X ^ (g.map (IsLocalRing.residue A)).order.toNat - f) := by
+  set n := (g.map (IsLocalRing.residue A)).order.toNat with hn
+  constructor
+  · refine (Polynomial.degree_sub_lt ?_ (Polynomial.monic_X_pow n).ne_zero ?_).trans_eq (by simpa)
+    · simp_rw [H.degree_eq_coe_lift_order_map, Polynomial.degree_X_pow, n,
+        ENat.lift_eq_toNat_of_lt_top]
+    · rw [(Polynomial.monic_X_pow n).leadingCoeff, H.isDistinguishedAt.monic.leadingCoeff]
+  · simp_rw [H.eq_mul, mul_assoc, IsUnit.mul_val_inv, mul_one, Polynomial.coe_sub,
+      Polynomial.coe_pow, Polynomial.coe_X, add_sub_cancel]
+
+section IsAdicComplete
+
+variable [IsAdicComplete (IsLocalRing.maximalIdeal A) A] {g : A⟦X⟧}
+  (hg : g.map (IsLocalRing.residue A) ≠ 0)
```

Comment:
I would recopy `hg` in the lemmas below to make them easier to understand alone.

Reply:
OK.